### PR TITLE
Updates for maplibre maps

### PR DESCRIPTION
**Pull request by LiamConnors**
- Update docs
- Add new px functions

### CODE REVIEW

File: doc/python/scattermapbox.md
Comment (archmoj) [inline doc/python/scattermapbox.md:305]: Do you want this part to describe new maps or mapbox?

**Comment by archmoj** [comment]
Great PR. Thanks @LiamConnors for all this. :medal_sports: :trophy: :star2: 
I suggest we set status of this PR to `on hold` as we did for https://github.com/plotly/plotly.js/pull/7015 and don't merge it.
That way we could simply look at the main differences between the two.

Then you could duplicate your branch, rename all the docs files that has `mapbox` in their name to `map` and then bring back the `mapbox` docs from master. That way we have docs for both mapbox and new maps.
Thank you!

**Comment by archmoj** [comment]
Here is a list of doc files that has `mapbox` in their name: 
```
./doc/python/mapbox-county-choropleth.md
./doc/python/mapbox-layers.md
./doc/python/filled-area-on-mapbox.md
./doc/python/mapbox-density-heatmaps.md
./doc/python/lines-on-mapbox.md
./doc/python/scattermapbox.md
./doc/python/hexbin-mapbox.md
```

**Comment by LiamConnors** [comment]
> Here is a list of doc files that has `mapbox` in their name:
> 
> ```
> ./doc/python/mapbox-county-choropleth.md
> ./doc/python/mapbox-layers.md
> ./doc/python/filled-area-on-mapbox.md
> ./doc/python/mapbox-density-heatmaps.md
> ./doc/python/lines-on-mapbox.md
> ./doc/python/scattermapbox.md
> ./doc/python/hexbin-mapbox.md
> ```

I think leaving the file names as is is okay for now as they are not displayed in the docs and it makes it easier for anyone reviewing this PR to see what's changed

**Comment by archmoj** [comment]
Yes we could keep mapbox names in this PR. But for the other one we want to merge those should be renamed as we need to keep doc for mapbox traces.

**Comment by LiamConnors** [comment]
> Yes we could keep mapbox names in this PR. But for the other one we want to merge those should be renamed as we need to keep doc for mapbox traces.

This PR keeps some mapbox examples. Like here https://github.com/plotly/plotly.py/pull/4706/files#diff-aeb7417e2580482c1fb42c4b976c4e0beb113fcb7333efee59c57bc06d3e04e6R138-R161
But I'm not sure we need to keep all examples written using Mapbox traces and the main focus of the pages should now be the new traces.
What do you think @ndrezn @emilykl

**Comment by archmoj** [comment]
IMHO, for a while we need to keep mapbox examples until they could use the next plotly.py version in their systems.
@ndrezn @emilykl @gvwilson

**Comment by archmoj** [comment]
Also having two versions (mapbox and new map) on two doc files, could help everyone see the difference between the two.

**Comment by ndrezn** [comment]
We should keep some mapbox examples, but like in `plotly.py` I would expect a deprecation notice associated with them. With the major version where we fully drop support for Mapbox I would also expect we drop these examples (or note that they're no longer available, perhaps for a brief period).

File: doc/python/choropleth-maps.md
Comment (ndrezn) [inline doc/python/choropleth-maps.md:36]: Maybe we should have in all of these charts a banner like:
> These trace types replace the previous `_mapbox` trace types, which will be deprecated in a feature version of Plotly. The core difference is that these traces use Maplibre as the rendering service rather than Mapbox, which allows us to keep our chart types up to the latest standards.

Or something? 

IMHO with this text we might also post on the forums and say, "If you have any feedback or want to learn more, you can see this post on our forums."

cc @gvwilson ; what do you think?

Comment (gvwilson) [reply doc/python/choropleth-maps.md:36]: I like the first sentence ("These trace types..."), but rather than details about Maplibre vs Mapbox, put in a URL to a page like the one we did for maintenance badges so that we can update as questions come in or our decisions change w/out having to do another release?

**Comment by archmoj** [comment]
There is also `hexbin_mapbox` in `figure_factory` at `packages/python/plotly/plotly/figure_factory/_hexbin_mapbox.py`.
We should possibly add `hexbin_map`. No?

**Comment by gvwilson** [comment]
I'm OK with shipping most of the maps in this release and doing another with the small number of maps we didn't get to.

File: doc/python/scattermapbox.md
Comment (LiamConnors) [inline doc/python/scattermapbox.md:304]:
```suggestion
## Mapbox Maps

> Mapbox traces are deprecated and may be removed in a future version of Plotly.py. 
```

**Comment by LiamConnors** [comment]
> We should keep some mapbox examples, but like in `plotly.py` I would expect a deprecation notice associated with them. With the major version where we fully drop support for Mapbox I would also expect we drop these examples (and note in the new docs that these maps replace the previous Mapbox trace types with feature parity).

Agreed. Kept one example for each page and they have the following deprecation notice 
https://github.com/plotly/plotly.py/pull/4706/commits/b849b3836155cd24b94e5960afe32e1ff3e04eaf

It's also mentioned that the other traces are the new way to create maps

File: doc/python/mapbox-county-choropleth.md
Comment (gvwilson) [inline doc/python/mapbox-county-choropleth.md:80]: if "DataFrame" is a class name, should it be in code font? If it's not, should it be lower case?

File: doc/python/mapbox-county-choropleth.md
Comment (gvwilson) [inline doc/python/mapbox-county-choropleth.md:255]: See _the_ function reference (?)

File: doc/python/mapbox-density-heatmaps.md
Comment (gvwilson) [inline doc/python/mapbox-density-heatmaps.md:97]: why is "stamen" quoted? other technical jargon isn't quoted...

File: doc/python/scattermapbox.md
Comment (gvwilson) [inline doc/python/scattermapbox.md:204]: setting _the_ symbol attribute (and should "symbol" be in code font?)

**Review by gvwilson — APPROVED** [review]
minor comments but looks good

File: doc/python/county-choropleth.md
Comment (LiamConnors) [reply doc/python/county-choropleth.md:40]: Yes "county" is correct

File: doc/python/datashader.md
Comment (LiamConnors) [reply doc/python/datashader.md]: fixed the formatting. thanks

File: doc/python/filled-area-on-mapbox.md
Comment (LiamConnors) [reply doc/python/filled-area-on-mapbox.md:45]: yes, it should for consistency. updated

File: doc/python/map-configuration.md
Comment (LiamConnors) [inline doc/python/map-configuration.md:42]:
```suggestion
If your figure is created with a `px.scatter_map`, `px.scatter_mapbox`, `px.line_map`, `px.line_mapbox`, `px.choropleth_map`, `px.choropleth_mapbox`, `px.density_map`, or `px.density_mapbox` function or otherwise contains one or more traces of type `go.Scattermap`, `go.Scattermapbox`,`go.Choroplethmap`, `go.Choroplethmapbox`, `go.Densitymap`, or `go.Densitymapbox` the `layout.map` object in your figure contains configuration information for the map itself.
```

File: doc/python/mapbox-county-choropleth.md
Comment (LiamConnors) [reply doc/python/mapbox-county-choropleth.md:255]: Yeah, I think that would be more correct, but it's already written this way across all other pages, so I won't update it in this PR.

File: doc/python/filled-area-on-mapbox.md
Comment (LiamConnors) [reply doc/python/filled-area-on-mapbox.md]: good idea. updated

File: doc/python/lines-on-mapbox.md
Comment (ndrezn) [inline doc/python/lines-on-mapbox.md:1]: We should change the name of this file? Ditto `filled-area-on-mapbox.md`.

**Review by ndrezn — CHANGES_REQUESTED** [review]
Pre comment above -- I think we're just blocked by merging until there is a doc specific to "Migration" noting _exactly_ what to change in your code to update.

We would link to that page in the deprecation warning in `plotly.py`.

File: doc/python/mapbox-density-heatmaps.md
Comment (LiamConnors) [inline doc/python/mapbox-density-heatmaps.md:108]:
```suggestion
fig = px.density_mapbox(df, lat='Latitude', lon='Longitude', z='Magnitude', radius=10,
                        center=dict(lat=0, lon=180), zoom=0,
```

File: doc/python/mapbox-density-heatmaps.md
Comment (LiamConnors) [inline doc/python/mapbox-density-heatmaps.md:97]:
```suggestion
#### Stamen Terrain base map with Mapbox (Stadia Maps token needed): density heatmap with `plotly.express`

Some base maps require a token. To use "stamen" base maps, you'll need a [Stadia Maps](https://www.stadiamaps.com) token, which you can provide to the `mapbox_accesstoken` parameter on `fig.update_layout`. Here, we have the token saved in a file called `.mapbox_token`, load it in to the variable `token`, and then pass it to `mapbox_accesstoken`.
```

File: doc/python/lines-on-mapbox.md
Comment (LiamConnors) [reply doc/python/lines-on-mapbox.md:1]: True. I left as is for now to make it easier for anyone reviewing the PR to see what changed

File: doc/python/migrate-to-maplibre.md
Comment (ndrezn) [inline doc/python/migrate-to-maplibre.md:70]: I would like to verify that these styles look okay in a preview deploy. I did not test locally.

Comment (archmoj) [reply doc/python/migrate-to-maplibre.md:70]: I don't like using direct links to master. 
Instead you may use either a release tag or a commit hash so that in future if `mapbox` images are removed it would not impact this.

Comment (ndrezn) [reply doc/python/migrate-to-maplibre.md:70]: ✅ Good call

Comment (archmoj) [reply doc/python/migrate-to-maplibre.md:70]: Also see https://github.com/plotly/plotly.js/commit/c31bc9fac5dafd75ae525208f198c1adbf8c4c22 for styles in map subplots vs styles in mapbox subplots.

Comment (ndrezn) [reply doc/python/migrate-to-maplibre.md:70]: I'd prefer even to copy these images and host elsewhere as they're test artifacts.

File: doc/python/migrate-to-maplibre.md
Comment (archmoj) [inline doc/python/migrate-to-maplibre.md:56]: By redirecting mapbox to map, perhaps we would be able to handle this change (in most cases) without users have to do the change in their code.
```suggestion
As a result of removing Mapbox as the rendering engine, we're also removing the Mapbox branding from these trace names. This means that migrating from Mapbox traces to MapLibre traces may require some code changes in your projects.
```

File: doc/python/migrate-to-maplibre.md
Comment (archmoj) [inline doc/python/migrate-to-maplibre.md:64]: Do you want to add info about using `stamen` styles as well?
cc: @emilykl 
https://github.com/plotly/plotly.js/pull/7085

File: doc/python/migrate-to-maplibre.md
Comment (LiamConnors) [inline doc/python/migrate-to-maplibre.md:33]: not sure we need a redirect here.

File: doc/python/migrate-to-maplibre.md
Comment (ndrezn) [reply doc/python/migrate-to-maplibre.md:64]: Yes, but is that something we should mention in the migration guide or in a different docs page? TBH it seems like it'd be nice to have a dedicated "Styles" page? but that can wait until the dust settles on this project.

File: doc/python/migrate-to-maplibre.md
Comment (ndrezn) [reply doc/python/migrate-to-maplibre.md:56]: I think you're right on this, and chatted with @alexcjohnson about how we can handle this (maybe by leveraging good old `clean_layout` which was/is used for mapping `is_category`/`is_log` to `access_types`).

That said for the minor release, if users want to take advantage of MapLibre, they _do_ need to change their code, so this prose is accurate today even if we might modify it in a future major release.

File: doc/python/migrate-to-maplibre.md
Comment (emilykl) [inline doc/python/migrate-to-maplibre.md:63]:
```suggestion
Note that Mapbox API keys are no longer required for Plotly-provided styles, but using external styles in your Plotly maps remains supported with the existing API.
```

File: doc/python/migrate-to-maplibre.md
Comment (LiamConnors) [inline doc/python/migrate-to-maplibre.md:54]:
```suggestion
You can learn more about the motivations for this change in our [announcement post](https://plotly.com/blog/plotly-is-switching-to-maplibre/).
```

File: doc/python/migrate-to-maplibre.md
Comment (LiamConnors) [reply doc/python/migrate-to-maplibre.md:64]: There is information on it here https://github.com/plotly/plotly.py/blob/b0c45b38808f1bd6b137ccbef636594091d620e9/doc/python/mapbox-layers.md

File: doc/python/figure-structure.md
Comment (archmoj) [inline doc/python/figure-structure.md:184]: Is `/python/tile-map-layers/` correct? Or it should be mapbox?

File: doc/python/datashader.md
Comment (LiamConnors) [reply doc/python/datashader.md:65]: @archmoj did you install the requirements from here?
https://github.com/plotly/plotly.py/blob/master/doc/requirements.txt#L34

File: doc/python/lines-on-mapbox.md
Comment (archmoj) [inline doc/python/lines-on-mapbox.md:94]: This example required installation of `wget` i.e. to download 
```py
wget.download("https://plotly.github.io/datasets/ne_50m_rivers_lake_centerlines.zip")
```
Wondering if we could simplify that or alternatively put `wget` and other required modules e.g. `datashader`  in requirements?

File: doc/python/datashader.md
Comment (LiamConnors) [reply doc/python/datashader.md:65]: I got it running here:

image

Comment (archmoj) [reply doc/python/datashader.md:65]: > @archmoj did you install the requirements from here? https://github.com/plotly/plotly.py/blob/master/doc/requirements.txt#L34

I installed them from here: https://github.com/plotly/plotly.py/blob/master/contributing.md#install-requirements---non-windows

Comment (LiamConnors) [reply doc/python/datashader.md:65]: Ah okay. The docs have additional requirements mentioned here

https://github.com/plotly/plotly.py/tree/master/doc#introduction-structure-and-required-packages

File: doc/python/lines-on-mapbox.md
Comment (LiamConnors) [reply doc/python/lines-on-mapbox.md:94]: they are in the docs requirements here https://github.com/plotly/plotly.py/blob/master/doc/requirements.txt and those are used when the docs are built

File: doc/python/figure-structure.md
Comment (LiamConnors) [inline doc/python/figure-structure.md:184]:
```suggestion
Figures can include two different types of map subplots: [geo subplots for outline maps](/python/map-configuration/) and [tile-based maps](/python/tile-map-layers/). The following trace types support attributes named `geo` or `map`, whose values must refer to corresponding objects in the layout i.e. `geo="geo2"` etc.  Note that attributes such as `layout.geo2` and `layout.map` etc do not have to be explicitly defined, in which case default values will be inferred. Multiple traces of a compatible type can be placed on the same subplot.
```

File: doc/python/datashader.md
Comment (archmoj) [reply doc/python/datashader.md:65]: If I try to install these requirements I get errors.
I suggest we have another `optional-docs` here: https://github.com/plotly/plotly.py/blob/master/contributing.md#install-requirements---non-windows
and just list some of these that are missing there.
Sounds reasonable?

File: doc/python/axes.md
Comment (archmoj) [inline doc/python/axes.md:44]: @LiamConnors 
Is `/python/map-layers/` correct? Or we should use `mapbox`?

File: doc/python/datashader.md
Comment (archmoj) [reply doc/python/datashader.md:65]: That's a separate issue.
Please feel free to resolve it for now.
I was able to test the graphs.
Nice work :1st_place_medal: 
Thank you! @LiamConnors

Comment (LiamConnors) [reply doc/python/datashader.md:65]: Sure, let's look at making that easier. Could you post the errors you got and version of Python and I'll open a new issue?

**Review by archmoj — APPROVED** [review]
I have one remaining comment here: https://github.com/plotly/plotly.py/pull/4706#discussion_r1735021656

Other than that it looks great and works great too!

You may consider merging #4726 into this branch or target master after merging this PR.

File: doc/python/datashader.md
Comment (archmoj) [reply doc/python/datashader.md:65]: I opened #4732 related to this.
Resolving...

File: doc/python/migrate-to-maplibre.md
Comment (ndrezn) [reply doc/python/migrate-to-maplibre.md:69]: @LiamConnors what's the best place to host these images?

Comment (LiamConnors) [reply doc/python/migrate-to-maplibre.md:69]: Probably best place is add them here: https://github.com/plotly/graphing-library-docs/tree/master/all_static/images and then reference those in the same way as you're currently doing with the new links

File: doc/python/axes.md
Comment (LiamConnors) [inline doc/python/axes.md:44]:
```suggestion
- [Map axes](/python/tile-map-layers/). The axis object is [`go.layout.Map`](/python/reference/layout/map/)
```

Comment (LiamConnors) [reply doc/python/axes.md]: good catch. was a typo

**Pull request merged**
